typescript/sdk/src/token/read.ts

```diff
@@ -0,0 +1,138 @@
+import { ethers, providers } from 'ethers';
+
+import {
+  ERC20__factory,
+  HypERC20Collateral__factory,
+} from '@hyperlane-xyz/core';
+import { ERC20Metadata, ERC20RouterConfig } from '@hyperlane-xyz/sdk';
+import { Address } from '@hyperlane-xyz/utils';
+
+import { DEFAULT_CONTRACT_READ_CONCURRENCY } from '../consts/crud.js';
+import { EvmHookReader } from '../hook/read.js';
+import { EvmIsmReader } from '../ism/read.js';
+import { MultiProvider } from '../providers/MultiProvider.js';
+import { ChainName } from '../types.js';
+
+type WarpRouteBaseMetadata = Record<
+  | 'mailbox'
+  | 'owner'
+  | 'wrappedTokenAddress'
+  | 'hook'
+  | 'interchainSecurityModule',
+  string
+>;
+
+interface WarpRouteReader {
+  deriveWarpRouteConfig(address: Address): Promise<Partial<ERC20RouterConfig>>;
+}
+
+export class EvmERC20WarpRouterReader implements WarpRouteReader {
+  provider: providers.Provider;
+  evmHookReader: EvmHookReader;
+  evmIsmReader: EvmIsmReader;
+
+  constructor(
+    protected readonly multiProvider: MultiProvider,
+    protected readonly chain: ChainName,
+    protected readonly concurrency: number = DEFAULT_CONTRACT_READ_CONCURRENCY,
+  ) {
+    this.provider = this.multiProvider.getProvider(chain);
+    this.evmHookReader = new EvmHookReader(multiProvider, chain, concurrency);
+    this.evmIsmReader = new EvmIsmReader(multiProvider, chain, concurrency);
+  }
+
+  /**
+   * Derives the configuration for a Hyperlane ERC20 router contract at the given address.
+   *
+   * @param address - The address of the Hyperlane ERC20 router contract.
+   * @returns The configuration for the Hyperlane ERC20 router.
+   *
+   */
+  async deriveWarpRouteConfig(
+    address: Address,
+  ): Promise<Omit<ERC20RouterConfig, 'type' | 'gas'>> {
+    const fetchedBaseMetadata = await this.fetchBaseMetadata(address);
+    const fetchedTokenMetadata = await this.fetchTokenMetadata(
+      fetchedBaseMetadata.wrappedTokenAddress,
+    );
+
+    const results: Omit<ERC20RouterConfig, 'type' | 'gas'> = {
```

Comment:
these 2 fields are omitted because 1) `type` can't be derived from the contracts. 2) Not sure if we need to return 'gas' for now. iiuc, this value comes from `destinationGas()` look up

Reply:
make `Omit<ERC20RouterConfig, 'type' | 'gas'>` a new type e.g. `DerivedERC20WarpRouteConfig` a la https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/055558a2661d94f8a9779f850a25d879ed413378/typescript/sdk/src/ism/read.ts#L44-L46